### PR TITLE
[REVIEW] Adding support to read strings with delimiters

### PR DESCRIPTION
**Pull request by ogreen**
String columns can now be read in with quotation marks at the beginning and end of the string. The quotation marks can be kept or removed based on user preference.
Also, if the string within the quotation mark has a field delimiter, that delimiter is ignored and not does not terminate the field.

### CODE REVIEW

File: cudf/io.py [hunk 1]
```diff
@@ -42,6 +42,9 @@ def read_csv(filepath, lineterminator='\n',
     dtype : list of str or dict of {col: dtype}, default None
         List of data types in the same order of the column names
         or a dictionary with column_name:dtype (pandas style).
+    quoting : bool, default False
+        If False, any quotes will be removed from start & end of string fields
+        If True, any quotes on fields are included in nvstrings entries
```
Comment (harrism) [inline cudf/io.py]:
```suggestion
        If True, any quotes will be included in string fields
```
Suggest using parallel structure as shown between the False and True conditions to make behavior clearer. Also consider swapping the order of True and False.

File: libgdf/src/io/csv/csv-reader.cu [hunk 2]
```diff
@@ -126,7 +127,7 @@ gdf_error launch_dataTypeDetection(raw_csv_t * raw_csv, long row_offset, column_
 __global__ void countRecords(char *data, const char delim, const char terminator, long num_bytes, long num_bits, unsigned long long* num_records);
 __global__ void storeRecordStart(char *data, const char delim, const char terminator, long num_bytes, long num_bits, unsigned long long* num_records,unsigned long long* recStart) ;
 
-__global__ void convertCsvToGdf(char *csv,char delim, char terminator,	unsigned long long num_records, int num_columns,bool *parseCol,unsigned long long *recStart,gdf_dtype *dtype,void **gdf_data,gdf_valid_type **valid,string_pair **str_cols,unsigned long long row_offset, long header_row,bool dayfirst,unsigned long long *num_valid);
+__global__ void convertCsvToGdf(char *csv,char delim, char terminator,	unsigned long long num_records, int num_columns,bool *parseCol,unsigned long long *recStart,gdf_dtype *dtype,void **gdf_data,gdf_valid_type **valid,string_pair **str_cols,unsigned long long row_offset, long header_row,bool dayfirst,unsigned long long *num_valid, bool keepQuotes);
```
Comment (harrism) [inline libgdf/src/io/csv/csv-reader.cu]:
```suggestion
__global__ void convertCsvToGdf(char *csv, char delim, char terminator, unsigned long long num_records, int num_columns, bool *parseCol, unsigned long long *recStart, gdf_dtype *dtype, void **gdf_data, gdf_valid_type **valid, string_pair **str_cols, unsigned long long row_offset, long header_row, bool dayfirst, unsigned long long *num_valid, bool keepQuotes);
```

File: libgdf/src/io/csv/csv-reader.cu [hunk 3]
```diff
@@ -220,6 +221,8 @@ gdf_error read_csv(csv_read_arg *args)
 	raw_csv->num_actual_cols	= args->num_cols;
 	raw_csv->num_active_cols	= args->num_cols;
 	raw_csv->num_records		= 0;
+	raw_csv->keepQuotes			= args->quoting;
```
Comment (harrism) [inline libgdf/src/io/csv/csv-reader.cu]:
```suggestion
	raw_csv->keepQuotes		= args->quoting;
```

File: libgdf/src/io/csv/csv-reader.cu [hunk 4]
```diff
@@ -932,7 +936,8 @@ __global__ void convertCsvToGdf(
 		unsigned long long 			row_offset,
 		long 			header_row,
 		bool			dayfirst,
-		unsigned long long			*num_valid
+		unsigned long long			*num_valid,
```
Comment (harrism) [inline libgdf/src/io/csv/csv-reader.cu]: Please fix spacing so parameters line up (regardless of tab settings).

File: libgdf/src/io/csv/csv-reader.cu [hunk 5]
```diff
@@ -962,18 +967,36 @@ __global__ void convertCsvToGdf(
...
+			  else if ((pos-1) >= start && raw_csv[pos-1]=='\"' )
 				break;
 			}
 			else if (raw_csv[pos] == terminator){
+ 			  if(quotation==false)
+				break;
+			  else if ((pos-1) >= start && raw_csv[pos-1]=='\"' )
 				break;
 			}
             else if(raw_csv[pos] == '\r' &&  ((pos+1) < stop && raw_csv[pos+1]=='\n')){
```
Comment (harrism) [inline libgdf/src/io/csv/csv-reader.cu]: Please fix indentation.

File: libgdf/src/io/csv/csv-reader.cu [hunk 6]
```diff
@@ -962,18 +967,36 @@ __global__ void convertCsvToGdf(
...
-            	stop--;
+  			  if(quotation==false){
```
Comment (harrism) [inline libgdf/src/io/csv/csv-reader.cu]: Indentation.

File: cudf/io.py (hunk 1 above)
Comment (randerzander) [reply cudf/io.py]: I'm trying, too char-count concisely (<80 chars per line), to convey that any quotes at the start and end of the field, _if they exist_, will be included in the returned nvstring array, but if the field wasn't wrapped in quotes, quotes will not be added. Will take a second stab at this docstring.

I don't want to call them "string fields" until we have deeper integration of nvstrings into cuDF.

File: libgdf/include/gdf/cffi/io_types.h [hunk 7]
```diff
@@ -88,6 +88,8 @@ typedef struct {
 
   char			*encoding;					// the data encoding, NULL = UTF-8
 
+  bool			quoting;					// TRUE: keep the quotes in the text. FALSE: remove the quotes from the strings. Only for quotes starting and ending the strings.
```
[the remark this answers is not in the document]
Comment (BradReesWork) [reply libgdf/include/gdf/cffi/io_types.h]: What is missing are:
- user specified quote character
- handling double quotes

**Comment by harrism** [comment]
@ogreen @randerzander can you guys push this PR forward?

File: libgdf/include/gdf/cffi/io_types.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (j-ieong) [reply libgdf/include/gdf/cffi/io_types.h]: Added support for both in my latest commits

File: libgdf/src/io/csv/csv-reader.cu (hunk 6 above)
Comment (j-ieong) [reply libgdf/src/io/csv/csv-reader.cu]: All the indentation issues should be fixed

**Comment by j-ieong** [comment]
@ogreen graciously gave me push access to his repo, so we can re-use this PR.

@harrism: The indentation issues, public API nvstring reference, and False/True default value is fixed. Thanks!

File: python/cudf/io/csv.py [hunk 8]
```diff
@@ -143,12 +155,13 @@ def read_csv(filepath, lineterminator='\n',
 
 
 def read_csv_strings(filepath, lineterminator='\n',
+                     quotechar='\0', quoting=True, doublequote=True,
```
Comment (randerzander) [inline python/cudf/io/csv.py]: Same q as for normal read_csv

File: python/cudf/io/csv.py [hunk 9]
```diff
@@ -42,6 +43,14 @@ def read_csv(filepath, lineterminator='\n',
     dtype : list of str or dict of {col: dtype}, default None
         List of data types in the same order of the column names
         or a dictionary with column_name:dtype (pandas style).
+    quotechar : char, default '\0' (empty)
```
Comment (randerzander) [inline python/cudf/io/csv.py]: Per Pandas's API, shouldn't the quotechar default be '"' instead of empty?

File: python/cudf/io/csv.py [hunk 10]
```diff
@@ -143,12 +155,13 @@ def read_csv(filepath, lineterminator='\n',
...
                      delimiter=',', sep=None, delim_whitespace=False,
                      skipinitialspace=False, names=None, dtype=None,
                      skipfooter=0, skiprows=0, dayfirst=False):
 
     import nvstrings
-    from .series import Series
+    from cudf.dataframe.series import Series
 
     """
     **Experimental**: This function provided only as an alpha way of providing
```
Comment (randerzander) [inline python/cudf/io/csv.py]: Looks like some previous commits got lost in the force push. I'll add a commit with the updated example docstring.

File: python/cudf/io/csv.py (hunk 9 above)
Comment (j-ieong) [reply python/cudf/io/csv.py]: You're right! I was only looking at the param description in the docs, where it didn't specify the default. Should've looked at the function declaration too. Fixed.

**Comment by randerzander** [comment]
@harrism not sure why I can't reply directly to your question

```
Since this is the public API for reading CSV into a DataFrame, the user might not know what "nvstrings" is, right?

How about "If True, quotes will be included in output DataFrame strings"? Also, you can save four characters by removing "any " from both the True and False cases, I think it's implied. :)
```

It's not returned in a cuDF DataFrame, rather as a list of cuDF Series and 'nvstrings', which are the Python class type for a column of strings. When the nvstrings repo & docs go live, we can cross link to the API for nvstrings.

**Comment by kkraus14** [comment]
This LGTM, @randerzander @harrism any other changes before merge?

**Comment by randerzander** [comment]
lgtm

**Comment by harrism** [comment]
Thanks for updates! All green, merging.

**Pull request merged**